File: backend-ruby/app/controllers/api/v1/nguoi_dung_controller.rb
class Api::V1::NguoiDungController < ApplicationController
  before_action :set_nguoi_dung, only: [:show, :update, :destroy]

  def index
    records = NguoiDung.includes(:vai_tro, :sinh_vien).order(NguoiDungID: :asc)
    render json: records.map { |record| serialize_user(record) }, status: :ok
  end

  def show
    render json: serialize_user(@nguoi_dung), status: :ok
  end

  def create
    record = NguoiDung.new(nguoi_dung_params.except(:MatKhau, :MaSinhVien, :MaDinhDanh))
    record.MatKhauHash = normalize_password(params.dig(:nguoi_dung, :MatKhau) || record.MatKhauHash)
    record.TrangThai = truthy_status(params.dig(:nguoi_dung, :TrangThai))

    if record.save
      sync_student_code(record, params.dig(:nguoi_dung, :MaSinhVien) || params.dig(:nguoi_dung, :MaDinhDanh))
      render json: serialize_user(record.reload), status: :created
    else
      render json: { errors: record.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Email already exists." }, status: :conflict
  end

  def update
    payload = nguoi_dung_params
    attrs = payload.except(:MatKhau, :MaSinhVien, :MaDinhDanh)
    attrs[:MatKhauHash] = normalize_password(payload[:MatKhau]) if payload[:MatKhau].present?
    attrs[:TrangThai] = truthy_status(payload[:TrangThai]) if payload.key?(:TrangThai)
    attrs[:CapNhatLuc] = Time.current

    if @nguoi_dung.update(attrs)
      sync_student_code(@nguoi_dung, payload[:MaSinhVien] || payload[:MaDinhDanh])
      render json: serialize_user(@nguoi_dung.reload), status: :ok
    else
      render json: { errors: @nguoi_dung.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Email already exists." }, status: :conflict
  end

  def destroy
    @nguoi_dung.destroy
    render json: { message: "Deleted user with ID #{@nguoi_dung.NguoiDungID}." }, status: :ok
  rescue ActiveRecord::InvalidForeignKey
    render json: { error: "Cannot delete user with ID #{@nguoi_dung.NguoiDungID} because the user is referenced by exam data." }, status: :conflict
  end

  private

  def set_nguoi_dung
    @nguoi_dung = NguoiDung.includes(:vai_tro, :sinh_vien).find_by(NguoiDungID: params[:id])
    unless @nguoi_dung
      render json: { error: "User with ID #{params[:id]} was not found." }, status: :not_found
    end
  end

  def nguoi_dung_params
    params.require(:nguoi_dung).permit(:Email, :HoTen, :MatKhau, :MatKhauHash, :VaiTroID, :TrangThai, :MaSinhVien, :MaDinhDanh)
  end

  def normalize_password(value)
    text = value.to_s.strip
    return "" if text.blank?
    text.start_with?("hashed_") ? text : "hashed_#{text}"
  end

  def truthy_status(value)
    return true if value.nil?
    value == true || value.to_s == "1" || value.to_s.downcase == "true"
  end

  def sync_student_code(user, code)
    text = code.to_s.strip
    return unless user.VaiTroID.to_i == 4 && text.present?

    student = user.sinh_vien || SinhVien.find_or_initialize_by(MaSinhVien: text)
    student.NguoiDungID = user.NguoiDungID
    student.MaSinhVien = text
    student.HoTen = user.HoTen
    student.Email = user.Email
    student.TrangThai = user.TrangThai
    student.save!
  end

  def serialize_user(user)
    {
      NguoiDungID: user.NguoiDungID,
      Email: user.Email,
      MatKhauHash: user.MatKhauHash,
      HoTen: user.HoTen,
      VaiTroID: user.VaiTroID,
      TenVaiTro: user.vai_tro&.TenVaiTro,
      TrangThai: user.TrangThai ? 1 : 0,
      MaSinhVien: user.sinh_vien&.MaSinhVien,
      TaoLuc: user.TaoLuc,
      CapNhatLuc: user.CapNhatLuc
    }
  end
end

File: backend-csharp/Controllers/NguoiDungController.cs
using System.Text.Json;
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/nguoi_dung")]
public class NguoiDungController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpGet]
    public async Task<IActionResult> Index()
    {
        List<NguoiDung> records = await dbContext.NguoiDung
            .Include(record => record.VaiTro)
            .Include(record => record.SinhVien)
            .OrderBy(record => record.NguoiDungID)
            .ToListAsync();

        return Ok(records.Select(SerializeUser));
    }

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        NguoiDung? record = await dbContext.NguoiDung
            .Include(user => user.VaiTro)
            .Include(user => user.SinhVien)
            .FirstOrDefaultAsync(user => user.NguoiDungID == id);

        return record is null ? NotFound(Error($"User with ID {id} was not found.")) : Ok(SerializeUser(record));
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] JsonElement body)
    {
        AccountPayload? payload = ReadBody<AccountPayload>(body, "nguoi_dung");
        if (payload is null)
        {
            return UnprocessableEntity(Errors("Invalid user payload."));
        }

        NguoiDung record = new()
        {
            Email = payload.Email.Trim().ToLowerInvariant(),
            HoTen = string.IsNullOrWhiteSpace(payload.HoTen) ? payload.Email : payload.HoTen.Trim(),
            MatKhauHash = NormalizePassword(payload.MatKhau ?? payload.MatKhauHash),
            VaiTroID = payload.VaiTroID <= 0 ? 4 : payload.VaiTroID,
            TrangThai = payload.StatusAsBool(),
            TaoLuc = DateTime.Now
        };

        dbContext.NguoiDung.Add(record);
        try
        {
            await dbContext.SaveChangesAsync();
            await SyncStudentCode(record, payload.MaSinhVien ?? payload.MaDinhDanh);
            await dbContext.Entry(record).Reference(user => user.VaiTro).LoadAsync();
            await dbContext.Entry(record).Reference(user => user.SinhVien).LoadAsync();
            return CreatedAtAction(nameof(Show), new { id = record.NguoiDungID }, SerializeUser(record));
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Email or generated student code already exists."));
        }
    }

    [HttpPatch("{id:int}")]
    [HttpPut("{id:int}")]
    public async Task<IActionResult> Update(int id, [FromBody] JsonElement body)
    {
        NguoiDung? record = await dbContext.NguoiDung.Include(user => user.SinhVien).FirstOrDefaultAsync(user => user.NguoiDungID == id);
        if (record is null)
        {
            return NotFound(Error($"User with ID {id} was not found."));
        }

        AccountPayload? payload = ReadBody<AccountPayload>(body, "nguoi_dung");
        if (payload is null)
        {
            return UnprocessableEntity(Errors("Invalid user payload."));
        }

        record.Email = payload.Email.Trim().ToLowerInvariant();
        record.HoTen = string.IsNullOrWhiteSpace(payload.HoTen) ? payload.Email : payload.HoTen.Trim();
        record.VaiTroID = payload.VaiTroID <= 0 ? record.VaiTroID : payload.VaiTroID;
        record.TrangThai = payload.StatusAsBool(record.TrangThai);
        record.CapNhatLuc = DateTime.Now;
        if (!string.IsNullOrWhiteSpace(payload.MatKhau))
        {
            record.MatKhauHash = NormalizePassword(payload.MatKhau);
        }

        try
        {
            await dbContext.SaveChangesAsync();
            await SyncStudentCode(record, payload.MaSinhVien ?? payload.MaDinhDanh);
            await dbContext.Entry(record).Reference(user => user.VaiTro).LoadAsync();
            await dbContext.Entry(record).Reference(user => user.SinhVien).LoadAsync();
            return Ok(SerializeUser(record));
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Email or generated student code already exists."));
        }
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Destroy(int id)
    {
        NguoiDung? record = await dbContext.NguoiDung.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error($"User with ID {id} was not found."));
        }

        dbContext.NguoiDung.Remove(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return Ok(new { message = $"Deleted user with ID {record.NguoiDungID}." });
        }
        catch (DbUpdateException)
        {
            return Conflict(Error($"Cannot delete user with ID {record.NguoiDungID} because the user is referenced by exam data."));
        }
    }

    internal static object SerializeUser(NguoiDung user) => new
    {
        user.NguoiDungID,
        user.Email,
        user.MatKhauHash,
        user.HoTen,
        user.VaiTroID,
        TenVaiTro = user.VaiTro?.TenVaiTro,
        TrangThai = user.TrangThai ? 1 : 0,
        MaSinhVien = user.SinhVien?.MaSinhVien,
        user.TaoLuc,
        user.CapNhatLuc
    };

    internal static string NormalizePassword(string? value)
    {
        string text = (value ?? string.Empty).Trim();
        return text.StartsWith("hashed_", StringComparison.Ordinal) ? text : $"hashed_{text}";
    }

    private async Task SyncStudentCode(NguoiDung user, string? code)
    {
        string text = (code ?? string.Empty).Trim();
        if (user.VaiTroID != 4 || string.IsNullOrWhiteSpace(text))
        {
            return;
        }

        SinhVien? student = await dbContext.SinhVien.FirstOrDefaultAsync(record => record.NguoiDungID == user.NguoiDungID)
            ?? await dbContext.SinhVien.FirstOrDefaultAsync(record => record.MaSinhVien == text);

        if (student is null)
        {
            student = new SinhVien { MaSinhVien = text, TaoLuc = DateTime.Now };
            dbContext.SinhVien.Add(student);
        }

        student.NguoiDungID = user.NguoiDungID;
        student.MaSinhVien = text;
        student.HoTen = user.HoTen;
        student.Email = user.Email;
        student.TrangThai = user.TrangThai;
        student.CapNhatLuc = DateTime.Now;
        await dbContext.SaveChangesAsync();
    }

    public sealed class AccountPayload
    {
        public string Email { get; set; } = string.Empty;
        public string HoTen { get; set; } = string.Empty;
        public string? MatKhau { get; set; }
        public string? MatKhauHash { get; set; }
        public int VaiTroID { get; set; } = 4;
        public JsonElement TrangThai { get; set; }
        public string? MaSinhVien { get; set; }
        public string? MaDinhDanh { get; set; }

        public bool StatusAsBool(bool fallback = true)
        {
            return TrangThai.ValueKind switch
            {
                JsonValueKind.False => false,
                JsonValueKind.True => true,
                JsonValueKind.Number => TrangThai.GetInt32() == 1,
                JsonValueKind.String => TrangThai.GetString() == "1" || string.Equals(TrangThai.GetString(), "true", StringComparison.OrdinalIgnoreCase),
                _ => fallback
            };
        }
    }
}
